golem-worker-executor-base/src/services/worker_fork.rs

```diff
@@ -0,0 +1,178 @@
+impl<Ctx: WorkerCtx, Svcs: HasAll<Ctx> + Send + Sync + 'static> WorkerFork
+    for DefaultWorkerFork<Ctx, Svcs>
+{
+    async fn fork(
+        &self,
+        source_worker_id: &OwnedWorkerId,
+        target_worker_id: &WorkerId,
+        oplog_index_cut_off: OplogIndex,
+    ) -> Result<(), GolemError> {
+        record_worker_call("fork");
+
+        let (owned_source_worker_id, owned_target_worker_id, source_worker_metadata) = self
+            .validate_worker_forking(
+                &source_worker_id.account_id,
+                &source_worker_id.worker_id,
+                target_worker_id,
+                oplog_index_cut_off,
+            )
+            .await?;
+
+        let target_worker_id = owned_target_worker_id.worker_id.clone();
+        let account_id = owned_target_worker_id.account_id.clone();
+
+        // Not sure if we should copy the metadata or not, or stick on to just default
+        let target_worker_metadata = WorkerMetadata {
+            worker_id: target_worker_id.clone(),
+            account_id,
+            env: source_worker_metadata.env.clone(),
+            args: source_worker_metadata.args.clone(),
+            created_at: Timestamp::now_utc(),
+            parent: source_worker_metadata.parent.clone(),
+            last_known_status: WorkerStatusRecord::default(),
+        };
+
+        let source_worker_instance = Worker::get_or_create_suspended(
+            &self.all,
+            &owned_source_worker_id,
+            Some(source_worker_metadata.args.clone()),
+            Some(source_worker_metadata.env.clone()),
+            None,
+            None,
+        )
+        .await?;
+
+        let source_oplog = source_worker_instance.oplog();
+
+        let initial_oplog_entry = source_oplog.read(OplogIndex::INITIAL).await;
+
+        // Update the oplog initial entry with the new worker
+        let target_initial_oplog_entry = initial_oplog_entry
+            .update_worker_id(&target_worker_id)
+            .ok_or(GolemError::unknown(
+                "Failed to update worker id in oplog entry",
+            ))?;
+
+        let new_oplog = self
+            .all
+            .oplog_service()
+            .create(
+                &owned_target_worker_id,
+                target_initial_oplog_entry,
+                target_worker_metadata,
+                Arc::new(RwLock::new(ExecutionStatus::Suspended {
+                    last_known_status: WorkerStatusRecord::default(), // default is idle, TODO: check if we need to update this or derive from the source last known status
+                    component_type: ComponentType::Durable, // Probably forking should fail if component type is ephemeral, or not?
+                    timestamp: Timestamp::now_utc(),
+                })),
+            )
+            .await;
+
+        let second_index = u64::from(OplogIndex::INITIAL.next());
+        let cut_off_index = u64::from(oplog_index_cut_off);
+
+        for index in second_index..=cut_off_index {
```

Comment:
I saw you created some oplog index range iterator, but not using it? Either use it or delete it or am I missing something?

Reply:
That was reusing an existing range functionality in `OplogService` somehow, which essentially doesn't require `get_or_suspended` if used

```
 async fn read_range(
        &self,
        owned_worker_id: &OwnedWorkerId,
        start_idx: OplogIndex,
        last_idx: OplogIndex,
    ) -> BTreeMap<OplogIndex, OplogEntry> {
```

Reply:
Ok, I forgot what I did 1.5 weeks back, I see what you mean. I will use the iterator I implemented :)